numpy/core/tests/test_umath.py

```diff
@@ -740,6 +741,81 @@ def test_float_remainder_corner_cases(self):
                 assert_(np.isnan(fmod), 'dt: %s, fmod: %s' % (dt, rem))
 
 
+class TestDivisionOverflows:
+    import operator
+    helper_lambdas = {
+        'zero': lambda dtype: 0,
+        'min': lambda dtype: np.iinfo(dtype).min,
+        'neg_min': lambda dtype: -np.iinfo(dtype).min,
+        'min-zero': lambda dtype: (np.iinfo(dtype).min, 0),
+        'neg_min-zero': lambda dtype: (-np.iinfo(dtype).min, 0),
+    }
+    overflow_results = {
+        np.remainder: helper_lambdas['zero'],
+        np.fmod: helper_lambdas['zero'],
+        operator.mod: helper_lambdas['zero'],
+        operator.floordiv: helper_lambdas['neg_min'],
+        np.floor_divide: helper_lambdas['neg_min'],
+        np.divmod: helper_lambdas['neg_min-zero']
+    }
+
+    @pytest.mark.parametrize("dividend_dtype",
+            np.sctypes['int'])
+    @pytest.mark.parametrize("divisor_dtype",
+            np.sctypes['int'])
+    @pytest.mark.parametrize("operation",
+            [np.remainder, np.fmod, np.divmod, np.floor_divide,
+             operator.mod, operator.floordiv])
+    @np.errstate(divide='raise', over='raise')
+    def test_overflows(self, dividend_dtype, divisor_dtype, operation):
+        # SIMD tries to perform the operation on as many elements as possible
+        # that is a multiple of the register's size. We resort to the
+        # default implementation for the leftover elements.
+        # We try to cover all paths here.
+        arrays = [np.array([np.iinfo(dividend_dtype).min]*i,
+                           dtype=dividend_dtype) for i in range(1, 129)]
+        divisor = np.array([-1], dtype=divisor_dtype)
+        # If dividend is a larger type than the divisor (`else` case),
+        # then, result will be a larger type than dividend and will not
+        # result in an overflow for `divmod` and `floor_divide`.
+        if np.dtype(dividend_dtype).itemsize >= np.dtype(
+                divisor_dtype).itemsize and operation in (
+                        np.divmod, np.floor_divide,
+                        TestDivisionOverflows.operator.floordiv):
+            with pytest.raises(
+                    FloatingPointError,
+                    match="overflow encountered in"):
+                result = operation(
+                            dividend_dtype(np.iinfo(dividend_dtype).min),
+                            divisor_dtype(-1)
+                        )
+
+            # Arrays
+            for a in arrays:
+                # In case of divmod, we need to flatten the result
+                # column first as we get a column vector of quotient and
+                # remainder and a normal flatten of the expected result.
+                with pytest.raises(FloatingPointError):
+                    result = np.array(operation(a, divisor)).flatten('f')
```

Comment:
OK, lets roll with this now, I think its good and can't think of some super elegant way thing.

I have one last request unfortunately.  Rather than using `pytest.raises` can we make this test use `pytest.warns(RuntimeWarning)` (maybe with a match for "over" or "divide", th old match should keep working).

The point is that if we set it to warn only, we can then add a check for the result after the `with pytest.warns` statement, since we do try to define a "correct" fill value for the invalid items.

Reply:
Ah yes, that makes sense. I added it back at https://github.com/numpy/numpy/pull/21507/commits/aaa4c911259bfaad8efcbac2051c5586bde5fe28